clippy_lints/src/hash_borrow_str_soundness.rs

```diff
@@ -0,0 +1,84 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::ty::implements_trait;
+use if_chain::if_chain;
+use rustc_hir::def::{DefKind, Res};
+use rustc_hir::{Item, ItemKind, Path, TraitRef};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::Ty;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::sym;
+// rustc_middle::ty::Term
+
+declare_clippy_lint! {
+    ///
+    ///  This lint is concerned with the soundness of types that implement traits
+    ///  Hash, Borrow<u8> and Borrow<str>:
+    ///  ```
+    /// impl Hash for MyType {
+    /// }
+    ///
+    /// impl Borrow<str> for MyType {
+    /// }
+    ///
+    /// impl Borrow<[u8]> for MyType {
+    /// }
+    /// ```
+    /// and the semantics of Borrow<T> (and BorrowMut). A type is free to borrow as several different
+    /// types.
+    ///
+    /// When providing implementations for Borrow<T>, one should consider whether the different
+    /// implementations should act as facets or representations of the underlying type. Generic code
+    /// typically uses Borrow<T> when it relies on the identical behavior of these additional trait
+    /// implementations. These traits will likely appear as additional trait bounds.
+    ///
+    /// In particular Eq, Ord and Hash must be equivalent for borrowed and owned values:
+    /// `x.borrow() == y.borrow()` should give the same result as `x == y`.
+    ///
+    /// It follows then that the following equivalence must hold:
+    ///  `hash(x) == hash((x as Borrow<[u8]>).borrow()) == hash((x as Borrow<str>).borrow())`
+    ///  Unfortunately it doesn't hold as `hash("abc") != hash("abc".as_bytes())`.
+    #[clippy::version = "1.75.0"]
+    pub HASH_BORROW_STR_SOUNDNESS,
+    correctness,
+    "default lint description"
+}
+
+declare_lint_pass!(HashBorrowStrSoundness => [HASH_BORROW_STR_SOUNDNESS]);
+
+impl LateLintPass<'_> for HashBorrowStrSoundness {
+    /// We are emitting this lint at the Hash impl of a type that implements all
+    /// three of Hash, Borrow<str> and Borrow<[u8]>.
+    ///
+    /// We check that we are in the Hash impl
+    /// Then we check that the type implements Borrow<str> and Borrow<[u8]>
+    fn check_item(&mut self, cx: &LateContext<'_>, item: &Item<'_>) {
+        if_chain! {
```

Comment:
We've been moving away from the `if_chain!` macro, over to let chains, since it provides better DX (rustfmt doesn't support macros) and we'd also like to test the feature in general. Can this be changed?

Reply:
Sure yeah. Already done! I just saw the commit removing `if_chain` and I was wondering why my build was, well, not building :D